docs/celestia-architecture/adr-009-cat-pool.md

```diff
@@ -0,0 +1,96 @@
+# ADR 009: Content addressable transaction pool
+
+## Changelog
+
+- 2023-01-11: Initial Draft (@cmwaters)
+
+## Context
+
+One of the criterias of success for Celestia as a reliable data availability layer is the ability to handle large transactional throughput. A component that plays a significant role in this is the mempool. It's purpose is to receive transactions from clients and broadcast them to all other nodes, eventually reaching the next block proposer who includes it in their block. Given Celestia's aggregator-like role whereby larger transactions, i.e. blobs, are expected to dominate network traffic, a content-addressable algorithm, common in many other [peer-to-peer file sharing protocols](https://en.wikipedia.org/wiki/InterPlanetary_File_System), could be far more beneficial than the current transaction-flooding protocol that Tendermint currently uses.
+
+This ADR describes the content addressable transaction protocol and through a comparative analysis with the existing gossip protocol, presents the case for it's adoption in Celestia.
+
+## Decision
+
+Use a content addressable transaction pool for disseminating transaction to nodes within the Celestia Network
+
+## Detailed Design
+
+The core idea is that each transaction can be referenced by a key, generated through a cryptographic hash function that reflects the content of the transaction. Nodes signal to one another which transactions they have via this key, and can request transactions they are missing through the key. This reduces the amount of duplicated transmission compared to a system which blindly sends received transactions to all other connected peers (as we will see in the consequences section).
+
+Full details on the exact protocol can be found in the [spec](../../mempool/cat/spec.md). Here, the document focuses on the main deciding points around the architecture:
+
+- It is assumed clients submit transactions to a single node in the network. Thus a node that receives a transaction through RPC will immediately broadcast it to all connected peers.
+- The new messages: `SeenTx` and `WantTx`, are broadcast over a new mempool channel `byte(0x31)` for backwards compatibility and to distinguish priorities. Nodes running the other mempools will not receive these messages and will be able to operate normally. Similarly, the interfaces used by Tendermint are not modified in any way, thus a node operator can easily switch between mempool versions.
+- Transaction gossiping takes priority over these "state" messages as to avoid situations where we receive a `SeenTx` and respond with a `WantTx` while the transaction is still queued in the nodes p2p buffer.
```

Comment:
I looked into this a little more and my assumption about how it works is different (predominantly because I was thinking about the v0.35 p2p stack and not v0.34). In the v0.35 p2p stack, inbound processing was handled in priority order. Messages with higher priority would be delivered to reactors before messages with lower priority. In v0.34, this is not the case. Only with outbound messages does priority make a difference. 

So in this example, If a `Tx` is sent by the reactor after a `SeenTx`, it is likely to actually be sent over the wire before the `SeenTx`. But the opposite is not true. If a node receives a `SeenTx` before a `Tx` over the wire, it will pass those messages to the reactor in that order. 

The situation I had in mind was: say State messages had a higher priority than transactions. This would mean that If a node sent me a transaction and directly afterwards another node sent me a `SeenTx` for that transaction, I would process the `SeenTx` first and as I don't have the transaction yet I would needlessly send a `WantTx` whereas the next message in the buffer is the `Tx` I just requested

Reply:
> The situation I had in mind was: say State messages had a higher priority than transactions. This would mean that If a node sent me a transaction and directly afterwards another node sent me a SeenTx for that transaction, I would process the SeenTx first and as I don't have the transaction yet I would needlessly send a WantTx whereas the next message in the buffer is the Tx I just requested

I see what you meant, thanks for the clarification! based on what you mentioned about the newer version of the P2P stack, it seems like updating the ADR about the priority of messages would be a good idea.